exchanges/kucoin/kucoin_websocket.go

```diff
@@ -916,396 +904,241 @@ func (ku *Kucoin) processMarketSnapshot(respData []byte) error {
-		marketTickerSnapshotChannel, marketTickerSnapshotForCurrencyChannel,
+		marketSnapshotChannel, marketSymbolSnapshotChannel,
 		marketOrderbookLevel2Channels, marketOrderbookLevel2to5Channel,
 		marketOrderbokLevel2To50Channel, marketCandlesChannel,
-		marketMatchChannel, indexPriceIndicatorChannel,
-		markPriceIndicatorChannel, marginFundingbookChangeChannel,
-		privateSpotTradeOrders, accountBalanceChannel,
-		marginPositionChannel, marginLoanChannel,
-		spotMarketAdvancedChannel:
-		return asset.Spot, nil
+		marketMatchChannel, indexPriceIndicatorChannel, markPriceIndicatorChannel,
+		privateSpotTradeOrders, accountBalanceChannel, spotMarketAdvancedChannel:
+		return asset.Spot
+	case marginFundingbookChangeChannel, marginPositionChannel, marginLoanChannel:
+		return asset.Margin
 	default:
-		return asset.Empty, errors.New("channel not supported")
+		return asset.Empty
 	}
 }
 
 // GenerateDefaultSubscriptions Adds default subscriptions to websocket.
-func (ku *Kucoin) GenerateDefaultSubscriptions() ([]stream.ChannelSubscription, error) {
-	channels := []string{}
-	if ku.CurrencyPairs.IsAssetEnabled(asset.Spot) == nil || ku.CurrencyPairs.IsAssetEnabled(asset.Margin) == nil {
-		channels = append(channels,
-			marketTickerChannel,
-			marketMatchChannel,
-			marketOrderbookLevel2Channels)
-	}
-	if ku.CurrencyPairs.IsAssetEnabled(asset.Margin) == nil {
-		channels = append(channels,
-			marginFundingbookChangeChannel)
-	}
-	if ku.CurrencyPairs.IsAssetEnabled(asset.Futures) == nil {
-		channels = append(channels,
-			futuresTickerV2Channel,
-			futuresOrderbookLevel2Depth50Channel)
-	}
-	var subscriptions []stream.ChannelSubscription
-	if ku.Websocket.CanUseAuthenticatedEndpoints() {
-		if ku.CurrencyPairs.IsAssetEnabled(asset.Spot) == nil {
-			channels = append(channels,
-				accountBalanceChannel,
-			)
-		}
-		if ku.CurrencyPairs.IsAssetEnabled(asset.Margin) == nil {
-			channels = append(channels,
-				marginPositionChannel,
-				marginLoanChannel,
-			)
-		}
-		if ku.CurrencyPairs.IsAssetEnabled(asset.Futures) == nil {
-			channels = append(channels,
-				// futures authenticated channels
-				futuresTradeOrdersBySymbolChannel,
-				futuresTradeOrderChannel,
-				futuresStopOrdersLifecycleEventChannel,
-				futuresAccountBalanceEventChannel)
+func (ku *Kucoin) GenerateDefaultSubscriptions() ([]subscription.Subscription, error) {
+	assetPairs := map[asset.Item]currency.Pairs{}
+	for _, a := range ku.GetAssetTypes(false) {
+		if p, err := ku.GetEnabledPairs(a); err == nil {
+			assetPairs[a] = p
+		} else {
+			assetPairs[a] = currency.Pairs{} // err is probably that Asset isn't enabled, but we don't care about errors of any type
 		}
 	}
-	var err error
-	var spotPairs currency.Pairs
-	if ku.CurrencyPairs.IsAssetEnabled(asset.Spot) == nil {
-		spotPairs, err = ku.GetEnabledPairs(asset.Spot)
-		if err != nil {
-			return nil, err
+	authed := ku.Websocket.CanUseAuthenticatedEndpoints()
+	var subscriptions = []subscription.Subscription{}
```

Comment:
There is no difference between `var subscriptions []subscriptions.Subscription` and `var subscriptions = []subscription.Subscription{}` in the compiler. No need to change it, but no need to add it in future either

Reply:
Yeah - just a brain fart. I'd never normally write this.

Reply:
Fixed 3a15f88e